File: ios/Core/Domain/Sources/CoreDomain/MusicLibraryRepository.swift
import Foundation

public protocol MusicLibraryRepository: Sendable {
    func observeAvailability() -> AsyncStream<MusicServiceAvailability>
    func observeOwnPlaylists() -> AsyncStream<[PlaylistSummary]>
    func observePlaylist(id: PlaylistId) -> AsyncStream<Playlist?>
    func observeSavedTracks() -> AsyncStream<SavedTracksResult>
    func observeTracks(refs: [TrackRef]) -> AsyncStream<[Track]>

    func refreshAvailability() async throws
    func refreshOwnPlaylists() async throws
    func refreshPlaylist(id: PlaylistId) async throws
    func refreshSavedTracks() async throws
    func refreshTracks(refs: [TrackRef]) async throws
}

File: android/core/domain/src/main/kotlin/com/mplayeraudio/core/domain/musiclibrary/MusicLibraryRepository.kt
package com.mplayeraudio.core.domain.musiclibrary

import kotlinx.coroutines.flow.Flow

interface MusicLibraryRepository {
    fun observeAvailability(): Flow<MusicServiceAvailability>

    fun observeOwnPlaylists(): Flow<List<PlaylistSummary>>

    fun observePlaylist(id: PlaylistId): Flow<Playlist?>

    fun observeSavedTracks(): Flow<SavedTracksResult>

    fun observeTracks(refs: List<TrackRef>): Flow<List<Track>>

    suspend fun refreshAvailability()

    suspend fun refreshOwnPlaylists()

    suspend fun refreshPlaylist(id: PlaylistId)

    suspend fun refreshSavedTracks()

    suspend fun refreshTracks(refs: List<TrackRef>)
}
